common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketCoreImpl.kt

```diff
@@ -0,0 +1,538 @@
+import io.homeassistant.companion.android.common.data.websocket.impl.entities.TemplateUpdatedEvent
+import io.homeassistant.companion.android.common.data.websocket.impl.entities.TriggerEvent
+import java.io.IOException
+import java.util.Collections
+import java.util.concurrent.atomic.AtomicLong
+import kotlin.coroutines.resumeWithException
+import kotlin.time.Duration.Companion.seconds
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.CoroutineStart
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ProducerScope
+import kotlinx.coroutines.channels.awaitClose
+import kotlinx.coroutines.channels.consumeEach
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.flow.shareIn
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withTimeoutOrNull
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.Response
+import okhttp3.WebSocket
+import okhttp3.WebSocketListener
+import okio.ByteString
+import okio.ByteString.Companion.toByteString
+import timber.log.Timber
+
+internal class WebSocketCoreImpl(
+    private val okHttpClient: OkHttpClient,
+    private val serverManager: ServerManager,
+    private val serverId: Int,
+) : WebSocketCore, WebSocketListener() {
+
+    private val coroutineExceptionHandler = CoroutineExceptionHandler { ctx, err -> Timber.e(err, "Uncaught exception in WebSocketRepositoryImpl") }
+
+    // TODO We never close the scope? (Should we tight it to some other scope?)
+    // TODO check SupervisorJob or Job?
+    private val ioScope = CoroutineScope(Dispatchers.IO + Job() + coroutineExceptionHandler)
+
+    // TODO any reason to not use a ConcurrentHashMap?
+    private val activeMessages = Collections.synchronizedMap(mutableMapOf<Long, WebSocketRequest>())
+
+    // Each message that we send needs a unique ID to match it to the answer
+    private val id = AtomicLong(1)
+    private var connection: WebSocket? = null
+    private var connectionState: WebSocketState? = null
+    private var connectionHaVersion: HomeAssistantVersion? = null
+    private val connectedMutex = Mutex()
+
+    // TODO Can't we just keep it in an AtomicBool?
+    private var connected = CompletableDeferred<Boolean>()
+    private val eventSubscriptionMutex = Mutex()
+
+    // TODO Clarify what is the goal here
+    private val messageQueue = Channel<Job>(capacity = Channel.UNLIMITED).apply {
```

Comment:
This is documented in the place it is used

https://github.com/home-assistant/android/blob/39869d7f4b56a57eb741121198f816c2f4867fa7/common/src/main/kotlin/io/homeassistant/companion/android/common/data/websocket/impl/WebSocketRepositoryImpl.kt#L873